local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	delete(v.procTable, pvName)
+}
+
+// DeleteBlockPV is meant to execute cleanup concurrently with other such cleanup processes.
+func (v *BlockCleanerImpl) DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	iters, expErr, exe, args, err := v.getBlockDeleteCommand(mountPath, config)
+	if err != nil {
+		return err
+	}
+
+	if v.isRunning(pv.Name) {
+		// Run in progress, nothing to do,
+		return nil
+	}
+	v.markRunning(pv.Name)
+	go v.asyncDeleteBlockPV(pv, iters, expErr, exe, args...)
+
+	return nil
+}
+
+func (v *BlockCleanerImpl) getBlockDeleteCommand(mountPath string, config common.MountConfig) (int, string, string,
+	[]string, error) {
+	iters := config.BlockConf.Iterations
+	if config.BlockConf.Iterations <= 0 {
+		// Atleast one iteration must be run.
+		iters = 1
+	}
+
+	switch config.BlockConf.Cleaner {
+
+	case common.BlockCleanupFastRest, "":
+		// mkfs -F /dev/sda && wipefs -a /dev/sda
+		mkfsCmd := fmt.Sprintf("%s -F %q && %s -a %q", MkfsCommand, mountPath, WipeComand, mountPath)
+		return 1, NoErrExpected, ShellCommand, []string{"-c", mkfsCmd}, nil
+
+	case common.BlockCleanupZero:
+		// dd if=/dev/zero of=/dev/sda bas=8096 status=progress
+		// Since we are trying to fill the block device, a healthy end to the execution would be the error
+		// message "No space left on device".
+		args := []string{"if=" + ZeroDev, "of=" + mountPath, "bs=" + DefaultBlockSize, "status=progress"}
+		return iters, DDFullErrExpected, DDCommand, args, nil
+
+	case common.BlockCleanupRandom:
+		// shred -n 3 /dev/sda # we defer to shred to run the iterations
+		return 1, NoErrExpected, ShredCommand, []string{"-vzf", "-n", strconv.Itoa(iters), mountPath}, nil
+
+	case common.BlockCleanupCustom:
+		// sh -C "custom_cmd /dev/sda"
+		return iters, NoErrExpected, ShellCommand, []string{"-c", config.BlockConf.CustomExec + " " + mountPath}, nil
+	}
+
+	return 0, NoErrExpected, "", []string{}, fmt.Errorf("Unknown BlockCleanerImpl type %q specified in config",
+		config.BlockConf.Cleaner)
+}
+
+func (v *BlockCleanerImpl) asyncDeleteBlockPV(pv *v1.PersistentVolume, iterations int, expectErrorOut string,
+	exe string, exeArgs ...string) error {
+	err := v.execAllIterations(pv.Name, iterations, expectErrorOut, exe, exeArgs...)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	// Remove API object
+	if err := v.APIUtil.DeletePV(pv.Name); err != nil {
+		glog.Error(err)
+		deletingLocalPVErr := fmt.Errorf("Error deleting block PV %q: %v", pv.Name, err.Error())
+		v.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeWarning, common.EventVolumeFailedDelete,
+			deletingLocalPVErr.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (v *BlockCleanerImpl) execAllIterations(pvName string, iterations int, expectErrOut string,
+	exe string, exeArgs ...string) error {
+
+	for i := 0; i < iterations; i++ {
+		if i > 0 {
+			glog.Infof("PV %s : Cleaner %s %s", pvName, exe, exeArgs)
```

Comment:
Cleaner pvname: %s cmd: %s

More consistent with other log lines on L210 L223

Reply:
ok

Reply:
same... made it more consistent.